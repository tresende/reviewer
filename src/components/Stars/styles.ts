import styled, { css } from 'styled-components'

export const Wrapper = styled.div`
  ${({ theme }) => css`
    display: flex;
    margin-right: ${theme.spacings.small};
  `}
`
