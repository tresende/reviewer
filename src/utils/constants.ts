const SCORE_LABEL = ['PÃ©ssimo', 'Ruim', 'Mediano', 'Bom', 'Perfeito']
const MIN_TEXT_LENGTH = 5
const MAX_RATING = 5
const DEFAULT_SCORE = 3
const ANIMATION_DURATION = 1000

export { SCORE_LABEL, MAX_RATING, ANIMATION_DURATION, MIN_TEXT_LENGTH, DEFAULT_SCORE }
