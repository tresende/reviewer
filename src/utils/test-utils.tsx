import { ReactElement } from 'react'
import { render, RenderOptions } from '@testing-library/react'

import { ThemeProvider } from 'styled-components'
import theme from 'styles/theme'
import { LoadingContext, LoadingContextData, LoadingContextDefaultValues } from 'hooks/use-loading'

type CustomRenderProps = {
  loadingProviderProps?: LoadingContextData
} & Omit<RenderOptions, 'queries'>

const customRender = (
  ui: ReactElement,
  { loadingProviderProps = LoadingContextDefaultValues, ...renderOptions }: CustomRenderProps = {}
) =>
  render(
    <ThemeProvider theme={theme}>
      <LoadingContext.Provider value={loadingProviderProps}>{ui}</LoadingContext.Provider>
    </ThemeProvider>,
    renderOptions
  )

export * from '@testing-library/react'
export { customRender as render }
