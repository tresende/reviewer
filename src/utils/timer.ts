export function timeout(ms?: number | undefined) {
  return new Promise((resolve) => setTimeout(resolve, ms))
}
