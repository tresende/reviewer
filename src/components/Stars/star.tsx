import { FaRegStar, FaStar } from 'react-icons/fa'

import theme from 'styles/theme'

import * as S from './styles'

export type StarProps = {
  onClick?: () => void
  outline?: boolean
  size: number
}

const Star = ({ onClick, outline, size }: StarProps) => {
  const props = { color: theme.colors.yellow, size }

  return <S.Button onClick={onClick}>{outline ? <FaRegStar {...props} /> : <FaStar {...props} />}</S.Button>
}

export default Star
