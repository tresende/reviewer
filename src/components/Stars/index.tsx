import { useState } from 'react'

import Star from './star'
import { SCORE_LABEL, MAX_RATING } from 'utils/constants'

import * as S from './styles'

export type StarsProps = {
  quantity?: number
  size?: number
  onClick?: (index: number) => void
}

const Stars = ({ quantity = 1, size = 28, onClick }: StarsProps) => {
  const [selected, setSelected] = useState(quantity)

  const handleClick = (index: number) => {
    if (!onClick) return
    setSelected(index)
    onClick(index)
  }

  const solid = [...Array(selected)]
  const outline = [...Array(MAX_RATING - selected)]

  return (
    <S.Wrapper title={SCORE_LABEL[selected - 1]}>
      {solid.map((_, index) => (
        <Star key={index} onClick={() => handleClick(index + 1)} size={size} />
      ))}
      {outline.map((_, index) => {
        const currentIndex = index + 1 + selected
        return <Star key={currentIndex} onClick={() => handleClick(currentIndex)} outline size={size} />
      })}
    </S.Wrapper>
  )
}

export default Stars
