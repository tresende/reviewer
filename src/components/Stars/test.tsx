import { render, screen } from 'utils/test-utils'
import Stars from '.'
import Star, { StarProps } from './star'

jest.mock('components/Stars/star', () => ({
  __esModule: true,
  default: function Mock(props: StarProps) {
    return <div onClick={props.onClick} data-testid={`star-${props.outline ? 'outline' : 'solid'}`} />
  }
}))

describe('<Stars />', () => {
  it('should render component correctly', () => {
    const { container } = render(<Stars />)
    expect(container.firstChild).toMatchSnapshot()
    expect(screen.getAllByTestId('star-solid')).toHaveLength(1)
  })

  it('should render 3 starts', () => {
    render(<Stars quantity={3} />)
    expect(screen.getAllByTestId('star-outline')).toHaveLength(2)
  })

  it('should execute callback on click in star', () => {
    const onClick = jest.fn()
    render(<Stars onClick={onClick} />)
    screen.getByTestId('star-solid').click()
    expect(onClick).toBeCalledTimes(1)
  })

  it('should ignore callback when onclick is null', () => {
    render(<Stars />)
    screen.getByTestId('star-solid').click()
  })

  describe('<Star />', () => {
    it('should render component correctly', () => {
      const { container } = render(<Star size={1} />)
      expect(container.firstChild).toMatchSnapshot()
    })
  })
})
