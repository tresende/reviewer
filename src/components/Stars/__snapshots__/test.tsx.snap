// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Stars /> <Star /> should render component correctly 1`] = `
<div
  data-testid="star-solid"
/>
`;

exports[`<Stars /> should render component correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-right: 2.4rem;
}

<div
  class="c0"
  title="PÃ©ssimo"
>
  <div
    data-testid="star-solid"
  />
  <div
    data-testid="star-outline"
  />
  <div
    data-testid="star-outline"
  />
  <div
    data-testid="star-outline"
  />
  <div
    data-testid="star-outline"
  />
</div>
`;
